src/shared/sleep-config.c

```diff
@@ -102,6 +107,179 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+/* If battery percentage capacity is less than equal to 5% return success */
+int battery_is_low(void) {
+        int r;
+
+        r = read_battery_capacity_percentage();
+        if (r == -ENOENT)
+               return false;
+        if (r < 0)
+               return r;
+
+        return r <= 5;
+        /* If battery current capacity percentage is equal to or less than 5%, return 1.
+         * We have not used battery capacity_level since value is set to full
+         * or Normal in case acpi is not working properly.
+         * In case of no battery 0 will be returned and system will be suspended for 1st cycle then hibernated */
+}
+
+/* battery percentage capacity fetched from capacity file and if in range 0-100 then returned */
+int read_battery_capacity_percentage(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r == -ENOENT)
+               return log_debug_errno(r, "/sys/class/power_supply/BAT0/capacity is unavailable. Assuming no battery exists: %m");
+               /* Handling case when no battery is present. Exception here file is manually deleted / corrupted */
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+
+        return battery_capacity;
+}
+
+/* Read system and battery identifier from specific location and generate hash of it */
+static int get_system_battery_identifier_hash(uint8_t ret[static SHA256_DIGEST_SIZE]) {
+        _cleanup_free_ char *bat_manufacturer = NULL, *bat_model_name = NULL, *bat_serial_number = NULL;
+        sd_id128_t machine_id, product_id;
+        struct sha256_ctx hash;
+        int r;
+
+        sha256_init_ctx(&hash);
+        r = read_one_line_file("/sys/class/power_supply/BAT0/manufacturer", &bat_manufacturer);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/manufacturer is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/manufacturer: %m");
+        else if (strlen(bat_manufacturer) == 0)
+               log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "/sys/class/power_supply/BAT0/manufacturer is empty: %m");
+        else
+               sha256_process_bytes(bat_manufacturer, strlen(bat_manufacturer) + 1, &hash);
```

Comment:
this is probably more a candidate for the siphash stuff, i.e. siphash24_compress_string()

Reply:
@poettering i have tried using this but keep getting segmentation fault
Jul 16 17:27:11 cere systemd-sleep[2921]: Attempting to estimate battery discharge rate after wakeup from 2min sleep
Jul 16 17:27:11 cere kernel: systemd-sleep[2921]: segfault at 28 ip 00007fb127106953 sp 00007ffd053440c0 error 4 in libsystemd-shared-251.so[7f>
Jul 16 17:27:11 cere kernel: Code: 83 ec 40 48 89 7d d8 48 89 75 d0 48 89 55 c8 48 8b 45 d8 48 89 45 e0 48 8b 55 e0 48 8b 45 d0 48 01 d0 48 89 >

i have pushed the changes. Not sure what am i missing here....

Reply:
Finally fixed. Thanks @bluca :)